README
Paranormal is an extremely customizable (pseudo-programmable)
visualization plugin for the X MultiMedia System, based (conceptually)
on Winamp's AVS plugin.  The goal of Paranormal is to be as flexible as
possible by providing a number of 'actuators' that perform various
tasks involved in generating an image.  These actuators can then be
combined to create a nearly unlimited number of effects.  These
combinations can be saved to a file to be brought back later or
distributed to your friends.

ChangeLog
05-13-01  Jamie Gennis <jgennis@mindspring.com>

	  * added a default preset to xmms.c
	  * updated README for 0.2.0 release
	  * bumped version number in configure.in up to 0.2.0
	  * converted container_simple and container_once to use
	  PN_ACTUATOR_*_FUNC macros
	  * added PN_ACTUATOR_*_FUNC macros to actuators.h
	  * implimented save button in actuator config dialog
	  * added save_preset() to presets.c

05-11-01  Jamie Gennis <jgennis@mindspring.com>

	  * implimented the 'load' button in the config dialog
	  * load_pn_rc in xmms.c no longer tries to load test.pna
	  * fixed config dialog's cancel button making the config dialog
	  inaccessable
	
05-10-01  Jamie Gennis <jgennis@mindspring.com>

	  * added container_cycle
	  * added SDL_INIT_TIMER to SDL_Init () call in paranormal.c
	
05-08-01  Jamie Gennis <jgennis@mindspring.com>

	  * added xform_bump_spin actuator
	  * fixed cmap_bwgradient not changing the high index
	  of the colormap
	
05-06-01  Jamie Gennis <jgennis@mindspring.com>

	  * fixed a mem leak in the container cleanup functions
	  * fixed a mem leak in xform cleanup functions
	  * decided to stop smoking crack before I code cleanup functions
	
05-02-01  Jamie Gennis <jgennis@mindspring.com>

	  * added changes to the wave functions for handling both channels
	  from Sam Vaughn <vaughns@purdue.edu>
	  * got cfg.c working! ;)
	
04-29-01  Jamie Gennis <jgennis@mindspring.com>

	  * renamed fade and blur actuators to general_fade and
	  general_blur, (respectively) ;)
	  * more stuffs in cfg.c (still non-functional)
	  * added container_remove() to containers.c
	  * added copy_actuator() to actuators.c
	
04-28-01  Jamie Gennis <jgennis@mindspring.com>

	  * added cfg.c to EXEC_SOURCES in Makefile.am
	  * wrote initial cfg.c (non-functional)
	  * changed xmms.c to call pn_configure() for pn_xmms_configure

04-26-01  Jamie Gennis <jgennis@mindspring.com>

	  * added 'paranormal' executable to Makefile.am
	  * wrote a primitive standalone.c
	  * added --enable-profiling option to configure.in
	  * speedup in apply_xform and blur_exec
	
04-25-01  Jamie Gennis <jgennis@mindspring.com>

	  * released paranormal 0.1.0
	  * added OPT_TYPE_BOOLEAN
	  * added sin_val[] and cos_val[] global precomputed arrays
	  * added wave_radial builtin by Josh Newton <newtonj@purdue.edu>
	  * added primitive wave_normalize and wave_smooth builtins
	  * fixed dates in ChangeLog (they were ahead 1 day) ;)
	  * updated configure.in for 0.1.0 release
	
04-24-01  Jamie Gennis <jgennis@mindspring.com>

	  * added resizability to paranormal.c
	  * cleanups in paranormal.c
	  * added xform_ripple builtin
	  * the nonfatal error reporting window is no longer
	  persistent between enables
	
04-23-01  Jamie Gennis <jgennis@mindspring.com>

	  * made the xform_vector struct a little smaller
	  * speedup in apply_xform
	  * blit_to_screen now uses SDL 'hardware' palette
	  * added -l option to pndoc for listing builtins
	  * fixed a bug in apply_xform that caused image artifacts
	  * speedup in blur_exec

04-22-01  Jamie Gennis <jgennis@mindspring.com>

	  * wrote initial xform code
	  * made the first ChangeLog entry ;)

pndoc.c
/* FIXME: Add stuff to this! */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>

#include "actuators.h"

/* needed for linking w/ builtins */
void pn_fatal_error (const char *fmt, ...) {}
void pn_error (const char *fmt, ...) {}
void pn_quit (void) {}

int
main (int argc, char **argv)
{
  int i, j;
  struct pn_actuator_desc *a;

  for (i=1; i<argc; i++)
    {
      if (*argv[i] == '-')
	{
	  switch (argv[i][1])
	    {
	    case 'l':
	      for (j=0; builtin_table[j]; j++)
		printf ("%s\n", builtin_table[j]->name);
	      break;
	    }
	  continue;
	}

      /* find the actuator */
      a = get_actuator_desc (argv[i]);
      if (! a)
	{
	  fprintf (stderr, "Unknown actuator: %s\n", argv[i]);
	  continue;
	}

      printf ("-------------------------------------\n");
      printf ("%s\n\n%s\n\n", a->name, a->doc);

      if (a->option_descs)
	{
	  printf ("Available options:\n");
	  for (j=0; a->option_descs[j].name; j++)
	    {
	      printf ("%s [", a->option_descs[j].name);
	      switch (a->option_descs[j].type)
		{
		case OPT_TYPE_INT:
		  printf ("INT");
		  break;
		case OPT_TYPE_FLOAT:
		  printf ("FLOAT");
		  break;
		case OPT_TYPE_STRING:
		  printf ("STRING");
		  break;
		case OPT_TYPE_COLOR:
		  printf ("COLOR");
		  break;
		case OPT_TYPE_COLOR_INDEX:
		  printf ("COLOR INDEX");
		  break;
		case OPT_TYPE_BOOLEAN:
		  printf ("BOOLEAN");
		  break;
		}
	      printf ("]: %s\n", a->option_descs[j].doc);
	    }
	}
      else
	printf ("No options available\n");
    }

  return 0;
}


actuators.c
#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <glib.h>

#include "actuators.h"
//#include "containers.h"

/* FIXME: container options override containees - fix this? */
/* FIXME: add actuator groups (by a group name string) */

/* FIXME: add support for copying containers' children (optionally) */
struct pn_actuator *
copy_actuator (const struct pn_actuator *a)
{
  struct pn_actuator *actuator;
  int i;    

  actuator = g_new (struct pn_actuator, 1);

  actuator->desc = a->desc;

  /* Make an options table */
  if (actuator->desc->option_descs)
    {
      /* count the options */
      for (i=0; actuator->desc->option_descs[i].name; i++);

      actuator->options = g_new (struct pn_actuator_option, i);
      for (i=0; actuator->desc->option_descs[i].name; i++)
	{
	  actuator->options[i].desc = &actuator->desc->option_descs[i];

	  /* copy the default options */
	  switch (actuator->desc->option_descs[i].type)
	    {
	    case OPT_TYPE_INT:
	    case OPT_TYPE_COLOR_INDEX:
	    case OPT_TYPE_FLOAT:
	    case OPT_TYPE_COLOR:
	    case OPT_TYPE_BOOLEAN:
	      memcpy (&actuator->options[i].val,
		      &a->options[i].val,
		      sizeof (union actuator_option_val));
	      break;
	    case OPT_TYPE_STRING:
	    }
	}

      /* the NULL option */
      actuator->options[i].desc = NULL;
    }
  else
    actuator->options = NULL;

  if (actuator->desc->init)
    actuator->desc->init (&actuator->data);

  return actuator;
}

struct pn_actuator_desc *
get_actuator_desc (const char *name)
{
  int i;

  for (i=0; builtin_table[i]; i++)
    if (! g_strcasecmp (name, builtin_table[i]->name))
      break;

  /* actuator not found */
  if (! builtin_table[i])
    return NULL;

  return builtin_table[i];
}

struct pn_actuator *
create_actuator (const char *name)
{
  int i;
  struct pn_actuator_desc *desc;
  struct pn_actuator *actuator;

  /* find the actuatoreration */
  desc = get_actuator_desc (name);

  if (! desc)
    return NULL;

  actuator = g_new (struct pn_actuator, 1);
  actuator->desc = desc;

  /* Make an options table */
  if (actuator->desc->option_descs)
    {
      /* count the options */
      for (i=0; actuator->desc->option_descs[i].name; i++);

      actuator->options = g_new (struct pn_actuator_option, i);
      for (i=0; actuator->desc->option_descs[i].name; i++)
	{
	  actuator->options[i].desc = &actuator->desc->option_descs[i];

	  /* copy the default options */
	  switch (actuator->desc->option_descs[i].type)
	    {
	    case OPT_TYPE_INT:
	    case OPT_TYPE_COLOR_INDEX:
	    case OPT_TYPE_FLOAT:
	    case OPT_TYPE_COLOR:
	    case OPT_TYPE_BOOLEAN:
	      memcpy (&actuator->options[i].val,
		      &actuator->desc->option_descs[i].default_val,
		      sizeof (union actuator_option_val));
	      break;
	    case OPT_TYPE_STRING:
	      /* NOTE: It's not realloc'ed so don't free it */
	      actuator->options[i].val.sval =
		actuator->desc->option_descs[i].default_val.sval;
	      break;
	    }
	}

      /* the NULL option */
      actuator->options[i].desc = NULL;
    }
  else
    actuator->options = NULL;

  if (actuator->desc->init)
    actuator->desc->init (&actuator->data);

  return actuator;
}

void
destroy_actuator (struct pn_actuator *actuator)
{
  int i;

  if (actuator->desc->cleanup)
    actuator->desc->cleanup (actuator->data);

  /* find any option val's that need to be freed */
  if (actuator->options)
    for (i=0; actuator->options[i].desc; i++)
      switch (actuator->options[i].desc->type)
	{
	case OPT_TYPE_INT:
	case OPT_TYPE_FLOAT:
	case OPT_TYPE_COLOR:
	case OPT_TYPE_COLOR_INDEX:
	case OPT_TYPE_BOOLEAN:
	  break;
	case OPT_TYPE_STRING:
	  if (actuator->options[i].val.sval
	      != actuator->options[i].desc->default_val.sval)
	    g_free ((char *)actuator->options[i].val.sval);
	}

  g_free (actuator->options);
  g_free (actuator);
}

void
exec_actuator (struct pn_actuator *actuator)
{
  g_assert (actuator);
  g_assert (actuator->desc);
  g_assert (actuator->desc->exec);
  actuator->desc->exec (actuator->options, actuator->data);
}



paranormal.c
/* FIXME: add fullscreen / screenshots */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <math.h>

#include <SDL.h>

#include "paranormal.h"
#include "actuators.h"

/* SDL stuffs */
static SDL_Surface *screen;

/* Globals */
struct pn_rc         *pn_rc;
struct pn_image_data *pn_image_data;
struct pn_sound_data *pn_sound_data;

/* Trig Pre-Computes */
float sin_val[360];
float cos_val[360];

/* **************** drawing doodads **************** */

static void
blit_to_screen (void)
{
  int j;

  SDL_LockSurface (screen);

  /* FIXME: add scaling support */

  SDL_SetPalette (screen, SDL_LOGPAL|SDL_PHYSPAL,
		  (SDL_Color*)pn_image_data->cmap, 0, 256);

  for (j=0; j<pn_image_data->height; j++)
      memcpy (screen->pixels + j*screen->pitch,
	      pn_image_data->surface[0] + j*pn_image_data->width,
	      pn_image_data->width);


  SDL_UnlockSurface (screen);

  SDL_UpdateRect (screen, 0, 0, 0, 0);
}

static void
resize_video (guint w, guint h)
{
  pn_image_data->width = w;
  pn_image_data->height = h;

  if (pn_image_data->surface[0])
    g_free (pn_image_data->surface[0]);
  if (pn_image_data->surface[1])
    g_free (pn_image_data->surface[1]);

  pn_image_data->surface[0] = g_malloc0 (w * h);
  pn_image_data->surface[1] = g_malloc0 (w * h);

  screen = SDL_SetVideoMode (w, h, 8, SDL_HWSURFACE |
			     SDL_HWPALETTE | SDL_RESIZABLE);
  if (! screen)
    pn_fatal_error ("Unable to create a new SDL window: %s",
		    SDL_GetError ());
}

static void
take_screenshot (void)
{
  char fname[32];
  struct stat buf;
  int i=0;

  do
    sprintf (fname, "pn_%05d.bmp", ++i);
  while (stat (fname, &buf) == 0);

  SDL_SaveBMP (screen, fname);
}

/* FIXME: This should resize the video to a user-set
   fullscreen res */
static void
toggle_fullscreen (void)
{
  SDL_WM_ToggleFullScreen (screen);  
  if (SDL_ShowCursor (SDL_QUERY) == SDL_ENABLE)
    SDL_ShowCursor (SDL_DISABLE);
  else
    SDL_ShowCursor (SDL_ENABLE);
}

/* **************** basic renderer management **************** */
void
pn_init (void)
{
  int i;

  pn_sound_data = g_new0 (struct pn_sound_data, 1);
  pn_image_data = g_new0 (struct pn_image_data, 1);

  if (SDL_Init (SDL_INIT_VIDEO | SDL_INIT_TIMER | SDL_INIT_NOPARACHUTE) < 0)
    pn_fatal_error ("Unable to initialize SDL: %s", SDL_GetError ());

  resize_video (640, 360);

  SDL_WM_SetCaption (PACKAGE " " VERSION, PACKAGE);

  for(i=0; i<360; i++)
    {
      sin_val[i] = sin(i*(M_PI/180.0));
      cos_val[i] = cos(i*(M_PI/180.0));
    }
}

void
pn_cleanup (void)
{
  SDL_FreeSurface (screen);
  SDL_Quit ();


  if (pn_image_data)
    {
      if (pn_image_data->surface[0])
	g_free (pn_image_data->surface[0]);
      if (pn_image_data->surface[1])
	g_free (pn_image_data->surface[1]);
      g_free (pn_image_data);
    }
  if (pn_sound_data)
    g_free (pn_sound_data);
}

/* Renders one frame and handles the SDL window */
void
pn_render (void)
{
  SDL_Event event;

  /* Handle window events */
  while (SDL_PollEvent (&event))
    {
      switch (event.type)
	{
	case SDL_QUIT:
	  pn_quit ();
	  g_assert_not_reached ();
	case SDL_KEYDOWN:
	  switch (event.key.keysym.sym)
	    {
	    case SDLK_ESCAPE:
	      pn_quit ();
	      g_assert_not_reached ();
	    case SDLK_RETURN:
	      if (event.key.keysym.mod & (KMOD_ALT | KMOD_META))
		toggle_fullscreen ();
	      break;
	    case SDLK_BACKQUOTE:
	      take_screenshot ();
	      break;
	    default:
	    }
	  break;
	case SDL_VIDEORESIZE:
	  resize_video (event.resize.w, event.resize.h);	  
	  break;
	}
    }

  if (pn_rc->actuator)
    {
      exec_actuator (pn_rc->actuator);
      blit_to_screen ();
    }
}

/* this MUST be called if a builtin's output is to surface[1]
   (by the builtin, after it is done) */
void
pn_swap_surfaces (void)
{
  guchar *tmp = pn_image_data->surface[0];
  pn_image_data->surface[0] = pn_image_data->surface[1];
  pn_image_data->surface[1] = tmp;
}












standalone.c
/* FIXME: This should eventually be a full-fledged standalone
   program... now it's just a tiny thing for debugging/profiling */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>

#include <glib.h>

#include "paranormal.h"
#include "actuators.h"
#include "presets.h"

static struct pn_actuator *actuator;

void
pn_set_rc (struct pn_rc *new_rc)
{
  /* FIXME: Write this! */
}

void
pn_error (const char *fmt, ...)
{
  va_list ap;

  va_start (ap, fmt);
  vfprintf (stderr, fmt, ap);
  va_end (ap);
  fprintf (stderr, "\n");
}

void
pn_fatal_error (const char *fmt, ...)
{
  va_list ap;

  va_start (ap, fmt);
  vfprintf (stderr, fmt, ap);
  va_end (ap);
  fprintf (stderr, "\n");

  pn_cleanup ();
  if (actuator)
    destroy_actuator (actuator);

  exit (-1);
}

void
pn_quit (void)
{
  pn_cleanup ();
  if (actuator)
    destroy_actuator (actuator);
  exit (0);
}

static void
generate_pcm_data (void)
{
  int j, i;

  for (j=0; j<2; j++)
    for (i=0; i<512; i++)
      pn_sound_data->pcm_data[j][i] = -5000.0 + (gint16) (10000.0 * rand ()/(RAND_MAX+1.0));
}

static void
generate_freq_data (void)
{
  int j, i;

  for (j=0; j<2; j++)
    for (i=0; i<256; i++)
      pn_sound_data->freq_data[j][i] = (gint16) (250.0 * rand ()/(RAND_MAX+1.0));
}

static void
load_pn_rc (void)
{
  if (! pn_rc)
    {
      pn_rc = g_new (struct pn_rc, 1);
      pn_rc->actuator = load_preset ("test.pna");
      if (! pn_rc->actuator)
	pn_error ("Unable to load the test actuator \"test.pna\"");
    }
}

int
main (int argc, char **argv)
{
  actuator = load_preset ("test.pna");
  if (! actuator)
    pn_fatal_error ("Unable to load the test actuator \"test.pna\"");

  load_pn_rc ();

  pn_init ();

  for (;;)
    {
      generate_pcm_data ();
      generate_freq_data ();
      pn_render ();
    }

  pn_cleanup ();
}


xmms.c
/* FIXME: issues with not uniniting variables between
   enables?  I wasn't too careful about that, but it
   seems to work fine.  If there are problems perhaps
   look for a bug there?
*/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <memory.h>
#include <math.h>
#include <setjmp.h>

#include <glib.h>
#include <gtk/gtk.h>
#include <xmms/plugin.h>
#include <SDL/SDL.h>
#include <SDL/SDL_thread.h>

#include "paranormal.h"
#include "actuators.h"
#include "presets.h"
#include "containers.h"

/* Error reporting dlg */
static GtkWidget *err_dialog;

/* Draw thread stuff */
/* FIXME: Do I need mutex for pn_done? */
static SDL_Thread *draw_thread = NULL;
static SDL_mutex *sound_data_mutex;
static SDL_mutex *config_mutex;

static gboolean pn_done = FALSE;
jmp_buf quit_jmp;
gboolean timeout_set = FALSE;
guint quit_timeout;

/* Sound stuffs */
static gboolean new_pcm_data = FALSE;
static gboolean new_freq_data = FALSE;
static gint16 tmp_pcm_data[2][512];
static gint16 tmp_freq_data[2][256];

/* XMMS interface */
static void pn_xmms_init (void);
static void pn_xmms_cleanup (void);
static void pn_xmms_about (void);
static void pn_xmms_configure (void);
static void pn_xmms_render_pcm (gint16 data[2][512]);
static void pn_xmms_render_freq (gint16 data[2][256]);

static VisPlugin pn_vp = 
{
  NULL,
  NULL,
  0,
  PACKAGE " " VERSION,
  2,
  2,
  pn_xmms_init,
  pn_xmms_cleanup,
  pn_xmms_about,
  pn_xmms_configure,
  NULL, /* disable_plugin */
  NULL, /* pn_xmms_playback_start */
  NULL, /* pn_xmms_playback_stop */
  pn_xmms_render_pcm,
  pn_xmms_render_freq
};

VisPlugin *
get_vplugin_info (void)
{
  return &pn_vp;
}

static void
load_pn_rc (void)
{
  struct pn_actuator *a, *b;

  if (! pn_rc)
      pn_rc = g_new0 (struct pn_rc, 1);

  /* load a default preset */
  pn_rc->actuator = create_actuator ("container_simple");
  if (! pn_rc->actuator) goto ugh;
  a = create_actuator ("container_once");
  if (! a) goto ugh;
  b = create_actuator ("cmap_bwgradient");
  if (! b) goto ugh;
  b->options[2].val.cval.r = b->options[2].val.cval.g = 0;
  container_add_actuator (a, b);
  container_add_actuator (pn_rc->actuator, a);
  a = create_actuator ("general_fade");
  if (! a) goto ugh;
  container_add_actuator (pn_rc->actuator, a);
  a = create_actuator ("xform_spin");
  if (! a) goto ugh;
  a->options[0].val.fval = -4.0;
  a->options[2].val.fval = 0.9;
  container_add_actuator (pn_rc->actuator, a);
  a = create_actuator ("wave_horizontal");
  if (! a) goto ugh;
  container_add_actuator (pn_rc->actuator, a);
  a = create_actuator ("general_blur");
  if (! a) goto ugh;
  container_add_actuator (pn_rc->actuator, a);

  return;

 ugh:
  if (pn_rc->actuator)
    destroy_actuator (pn_rc->actuator);
  pn_error ("Error loading default preset");
}

static int
draw_thread_fn (gpointer data)
{
  pn_init ();

  /* Used when pn_quit is called from this thread */
  if (setjmp (quit_jmp) != 0)
    pn_done = TRUE;

  while (! pn_done)
    {
      SDL_mutexP (sound_data_mutex);
      if (new_freq_data)
	{
	  memcpy (pn_sound_data->freq_data, tmp_freq_data,
		  sizeof (gint16) * 2 * 256);
	  new_freq_data = FALSE;
	}
      if (new_pcm_data)
	{
	  memcpy (pn_sound_data->pcm_data, tmp_pcm_data,
		  sizeof (gint16) * 2 * 512);
	  new_freq_data = FALSE;
	}
      SDL_mutexV (sound_data_mutex);
      SDL_mutexP (config_mutex);
      pn_render ();
      SDL_mutexV (config_mutex);
    }

  /* Just in case a pn_quit () was called in the loop */
/*    SDL_mutexV (sound_data_mutex); */

  pn_cleanup ();

  return 0;
}

/* Is there a better way to do this? this = messy
   It appears that calling disable_plugin () in some
   thread other than the one that called pn_xmms_init ()
   causes a seg fault :( */
static int
quit_timeout_fn (gpointer data)
{
  if (pn_done)
    {
      pn_vp.disable_plugin (&pn_vp);
      return FALSE;
    }

  return TRUE;
}

static void
pn_xmms_init (void)
{
  /* If it isn't already loaded, load the run control */
  load_pn_rc ();

  sound_data_mutex = SDL_CreateMutex ();
  config_mutex = SDL_CreateMutex ();
  if (! sound_data_mutex)
    pn_fatal_error ("Unable to create a new mutex: %s",
		    SDL_GetError ());

  pn_done = FALSE;
  draw_thread = SDL_CreateThread (draw_thread_fn, NULL);
  if (! draw_thread)
    pn_fatal_error ("Unable to create a new thread: %s",
		    SDL_GetError ());

  /* Add a gtk timeout to test for quits */
  quit_timeout = gtk_timeout_add (1000, quit_timeout_fn, NULL);
  timeout_set = TRUE;
}

static void
pn_xmms_cleanup (void)
{
  if (timeout_set)
    {
      gtk_timeout_remove (quit_timeout);
      timeout_set = FALSE;
    }

  if (draw_thread)
    {
      pn_done = TRUE;
      SDL_WaitThread (draw_thread, NULL);
      draw_thread = NULL;
    }

  if (sound_data_mutex)
    {
      SDL_DestroyMutex (sound_data_mutex);
      sound_data_mutex = NULL;
    }

  if (config_mutex)
    {
      SDL_DestroyMutex (config_mutex);
      config_mutex = NULL;
    }
}

static void
about_close_clicked(GtkWidget *w, GtkWidget **window)
{
	gtk_widget_destroy(*window);
	*window=NULL;
}

static void
about_closed(GtkWidget *w, GdkEvent *e, GtkWidget **window)
{
	about_close_clicked(w,window);
}

static void
pn_xmms_about (void)
{
  static GtkWidget *window=NULL;
  GtkWidget *vbox, *buttonbox, *close, *label;

  if(window)
    return;

  window=gtk_window_new(GTK_WINDOW_TOPLEVEL);
  gtk_window_set_title(GTK_WINDOW(window), PACKAGE " " VERSION);
  gtk_window_set_policy(GTK_WINDOW(window), FALSE, FALSE, FALSE);

  vbox=gtk_vbox_new(FALSE, 4);
  gtk_container_add(GTK_CONTAINER(window), vbox);
  gtk_container_set_border_width(GTK_CONTAINER(vbox), 8);
  gtk_widget_show(vbox);

  label=gtk_label_new("\n\
" PACKAGE " " VERSION "\n\n\
Copyright (C) 2001, Jamie Gennis. (jgennis@mindspring.com)\n\
\n\
This program is free software; you can redistribute it and/or modify\n\
it under the terms of the GNU General Public License as published by\n\
the Free Software Foundation; either version 2 of the License, or\n\
(at your option) any later version.\n\
\n\
This program is distributed in the hope that it will be useful,\n\
but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
GNU General Public License for more details.\n\
\n\
You should have received a copy of the GNU General Public License\n\
along with this program; if not, write to the Free Software\n\
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307\n\
USA");

  gtk_box_pack_start(GTK_BOX(vbox), label, FALSE, FALSE, 8);
  gtk_widget_show(label);
	
  buttonbox=gtk_hbutton_box_new();
  gtk_box_pack_end(GTK_BOX(vbox), buttonbox, FALSE, FALSE,8);
  gtk_widget_show(buttonbox);
	
  close=gtk_button_new_with_label("Close");
  GTK_WIDGET_SET_FLAGS(close, GTK_CAN_DEFAULT);
  gtk_window_set_default(GTK_WINDOW(window), close);
  gtk_hbutton_box_set_layout_default(GTK_BUTTONBOX_END);
  gtk_box_pack_end(GTK_BOX(buttonbox), close, FALSE, FALSE,8);
  gtk_widget_show(close);
	
  gtk_signal_connect(GTK_OBJECT(close), "clicked", GTK_SIGNAL_FUNC(about_close_clicked), &window);
  gtk_signal_connect(GTK_OBJECT(window), "delete-event", GTK_SIGNAL_FUNC(about_closed), &window);

  gtk_widget_show(window);
}

static void
pn_xmms_configure (void)
{
  load_pn_rc ();

  /* We should already have a GDK_THREADS_ENTER
     but we need to give it config_mutex */
  if (config_mutex)
    SDL_mutexP (config_mutex);

  pn_configure ();

  if (config_mutex)
    SDL_mutexV (config_mutex);
}

static void
pn_xmms_render_pcm (gint16 data[2][512])
{
  SDL_mutexP (sound_data_mutex);
  memcpy (tmp_pcm_data, data, sizeof (gint16) * 2 * 512);
  new_pcm_data = TRUE;
  SDL_mutexV (sound_data_mutex);
}

static void
pn_xmms_render_freq (gint16 data[2][256])
{
  SDL_mutexP (sound_data_mutex);
  memcpy (tmp_freq_data, data, sizeof (gint16) * 2 * 256);
  new_freq_data = TRUE;
  SDL_mutexV (sound_data_mutex);
}

/* **************** paranormal.h stuff **************** */

void
pn_set_rc (struct pn_rc *new_rc)
{
  if (config_mutex)
    SDL_mutexP (config_mutex);

  if (pn_rc->actuator)
    destroy_actuator (pn_rc->actuator);
  pn_rc->actuator = new_rc->actuator;

  if (config_mutex)
    SDL_mutexV (config_mutex);
}

void
pn_fatal_error (const char *fmt, ...)
{
  char *errstr;
  va_list ap;
  GtkWidget *dialog;
  GtkWidget *close, *label;

  /* Don't wanna try to lock GDK if we already have it */
  if (draw_thread && SDL_ThreadID () == SDL_GetThreadID (draw_thread))
    GDK_THREADS_ENTER ();

  /* now report the error... */
  va_start (ap, fmt);
  errstr = g_strdup_vprintf (fmt, ap);
  va_end (ap);

  dialog=gtk_dialog_new();
  gtk_window_set_title(GTK_WINDOW(dialog), "Error - " PACKAGE " " VERSION);
  gtk_container_border_width (GTK_CONTAINER (dialog), 8);

  label=gtk_label_new(errstr);
  fprintf (stderr, "%s\n", errstr);
  g_free (errstr);

  close = gtk_button_new_with_label ("Close");
  gtk_signal_connect_object (GTK_OBJECT (close), "clicked",
			     GTK_SIGNAL_FUNC (gtk_widget_destroy),
			     GTK_OBJECT (dialog));

  gtk_box_pack_start (GTK_BOX (GTK_DIALOG (dialog)->vbox), label, FALSE,
		      FALSE, 0);
  gtk_box_pack_start (GTK_BOX (GTK_DIALOG (dialog)->action_area), close,
		      FALSE, FALSE, 0);
  gtk_widget_show (label);
  gtk_widget_show (close);

  gtk_widget_show (dialog);
  gtk_widget_grab_focus (dialog);

  if (draw_thread && SDL_ThreadID () == SDL_GetThreadID (draw_thread))
    GDK_THREADS_LEAVE ();

  pn_quit ();
}


void
pn_error (const char *fmt, ...)
{
  char *errstr;
  va_list ap;
  static GtkWidget *text;

  /* now report the error... */
  va_start (ap, fmt);
  errstr = g_strdup_vprintf (fmt, ap);
  va_end (ap);
  fprintf (stderr, PACKAGE ": %s\n", errstr);

  /* This is the easiest way of making sure we don't
     get stuck trying to lock a mutex that this thread
     already owns since this fn can be called from either
     thread */
  if (draw_thread && SDL_ThreadID () == SDL_GetThreadID (draw_thread))
    GDK_THREADS_ENTER ();

  if (! err_dialog)
    {
      GtkWidget *close;

      err_dialog=gtk_dialog_new();
      gtk_window_set_title (GTK_WINDOW (err_dialog), "Error - " PACKAGE " " VERSION);
      gtk_window_set_policy (GTK_WINDOW (err_dialog), FALSE, FALSE, FALSE);
      gtk_widget_set_usize (err_dialog, 400, 200);
      gtk_container_border_width (GTK_CONTAINER (err_dialog), 8);

      text = gtk_text_new (NULL, NULL);
      gtk_text_set_editable (GTK_TEXT (text), FALSE);
      gtk_text_set_word_wrap (GTK_TEXT (text), TRUE);

      close = gtk_button_new_with_label ("Close");
      gtk_signal_connect_object (GTK_OBJECT (close), "clicked",
				 GTK_SIGNAL_FUNC (gtk_widget_hide),
				 GTK_OBJECT (err_dialog));
      gtk_signal_connect_object (GTK_OBJECT (err_dialog), "delete-event",
				 GTK_SIGNAL_FUNC (gtk_widget_hide),
				 GTK_OBJECT (err_dialog));

      gtk_box_pack_start (GTK_BOX (GTK_DIALOG (err_dialog)->vbox), text, FALSE,
			  FALSE, 0);
      gtk_box_pack_start (GTK_BOX (GTK_DIALOG (err_dialog)->action_area), close,
			  FALSE, FALSE, 0);
      gtk_widget_show (text);
      gtk_widget_show (close);
    }

  gtk_text_set_point (GTK_TEXT (text), gtk_text_get_length (GTK_TEXT (text)));
  gtk_text_insert (GTK_TEXT (text), NULL, NULL, NULL, errstr, -1);
  gtk_text_insert (GTK_TEXT (text), NULL, NULL, NULL, "\n", 1);
  g_free (errstr);

  gtk_widget_show (err_dialog);
  gtk_widget_grab_focus (err_dialog);

  if (draw_thread && SDL_ThreadID () == SDL_GetThreadID (draw_thread))
    GDK_THREADS_LEAVE ();
}


/* This is confusing...
   Don't call this from anywhere but the draw thread or
   the initialization xmms thread (ie NOT the xmms sound
   data functions) */
void
pn_quit (void)
{
  if (draw_thread && SDL_ThreadID () == SDL_GetThreadID (draw_thread))
    {
      /* We're in the draw thread so be careful */
      longjmp (quit_jmp, 1);
    }
  else
    {
      /* We're not in the draw thread, so don't sweat it...
	 addendum: looks like we have to bend over backwards (forwards?)
	 for xmms here too */
      pn_vp.disable_plugin (&pn_vp);
      while (1)
	gtk_main_iteration ();
    }
}
